packages/lib-classifier/src/store/RootStore.js

```diff
@@ -98,6 +98,12 @@ const RootStore = types
       onPatch(self, _onPatch)
     }
 
+    function refreshWorkflow(workflowSnapshot, subjectSetID, subjectID) {
+      self.workflows.setResources([workflowSnapshot])
+      self.workflows.selectWorkflow(workflowSnapshot.id, subjectSetID, subjectID)
```

Comment:
I think this promise needs to resolve with a new active workflow before starting a new classification. Otherwise the subject queue might be advanced for the previous workflow.

Reply:
@shaunanoordin this is why you're seeing stale subjects in the queue after a workflow change. `onSubjectAdvance()` is running before the workflow has changed.

Reply:
Oooh I might need to trace `onSubjectAdvance()` then, as I'm honestly not 100% sure of the sequence of triggers here.

One thing to note though is that if `onSubjectAdvance()` is the problem, it might not just be contained to the call in `refreshWorkflow()`. [My recent tests](https://github.com/zooniverse/front-end-monorepo/pull/2677#issuecomment-1021612402) have shown that that the problem (Issue 2) is extant in _master_ as well.